sqlite3 survey.db

In [7]:
import sqlite3
import pandas as pd

conexao = sqlite3.connect('/home/fernando/github/python-datascience/survey.db')


OperationalError: unable to open database file

In [77]:
survey = pd.read_sql_query("SELECT * FROM Survey", conexao)
survey.info()
survey.describe()
print(survey)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21 entries, 0 to 20
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   taken    21 non-null     int64  
 1   person   19 non-null     object 
 2   quant    21 non-null     object 
 3   reading  21 non-null     float64
dtypes: float64(1), int64(1), object(2)
memory usage: 800.0+ bytes
    taken person quant  reading
0     619   dyer   rad     9.82
1     619   dyer   sal     0.13
2     622   dyer   rad     7.80
3     622   dyer   sal     0.09
4     734     pb   rad     8.41
5     734   lake   sal     0.05
6     734     pb  temp   -21.50
7     735     pb   rad     7.22
8     735   None   sal     0.06
9     735   None  temp   -26.00
10    751     pb   rad     4.35
11    751     pb  temp   -18.50
12    751   lake   sal     0.10
13    752   lake   rad     2.19
14    752   lake   sal     0.09
15    752   lake  temp   -16.00
16    752    roe   sal    41.60
17    837   lake   rad     

In [78]:
pessoa = pd.read_sql_query("SELECT * FROM person", conexao)
pessoa.info()
pessoa.describe()
print(pessoa)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   id        5 non-null      object
 1   personal  5 non-null      object
 2   family    5 non-null      object
dtypes: object(3)
memory usage: 248.0+ bytes
         id   personal    family
0      dyer    William      Dyer
1        pb      Frank   Pabodie
2      lake   Anderson      Lake
3       roe  Valentina   Roerich
4  danforth      Frank  Danforth


In [79]:


dfsite = pd.read_sql_query("SELECT * FROM Site", conexao)
print(dfsite.describe())
print(dfsite)



             lat        long
count   3.000000    3.000000
mean  -48.623333 -126.230000
std     1.366797    2.619599
min   -49.850000 -128.570000
25%   -49.360000 -127.645000
50%   -48.870000 -126.720000
75%   -48.010000 -125.060000
max   -47.150000 -123.400000
    name    lat    long
0   DR-1 -49.85 -128.57
1   DR-3 -47.15 -126.72
2  MSK-4 -48.87 -123.40


In [80]:
dfvisited=pd.read_sql_query("SELECT max(id) FROM Visited;", conexao)
print(dfvisited)

dfvisited=pd.read_sql_query("SELECT * FROM Visited;", conexao)
print(dfvisited['id'].max())


   max(id)
0      844
844


In [81]:
# Selecionando colunas

dfPerson = pd.read_sql_query("SELECT family, personal FROM Person;", conexao)
print(dfPerson)
dfSurvey = pd.read_sql_query("SELECT * FROM Survey;", conexao)
print(dfSurvey)
print(dfPerson['family'])

     family   personal
0      Dyer    William
1   Pabodie      Frank
2      Lake   Anderson
3   Roerich  Valentina
4  Danforth      Frank
    taken person quant  reading
0     619   dyer   rad     9.82
1     619   dyer   sal     0.13
2     622   dyer   rad     7.80
3     622   dyer   sal     0.09
4     734     pb   rad     8.41
5     734   lake   sal     0.05
6     734     pb  temp   -21.50
7     735     pb   rad     7.22
8     735   None   sal     0.06
9     735   None  temp   -26.00
10    751     pb   rad     4.35
11    751     pb  temp   -18.50
12    751   lake   sal     0.10
13    752   lake   rad     2.19
14    752   lake   sal     0.09
15    752   lake  temp   -16.00
16    752    roe   sal    41.60
17    837   lake   rad     1.46
18    837   lake   sal     0.21
19    837    roe   sal    22.50
20    844    roe   rad    11.25
0        Dyer
1     Pabodie
2        Lake
3     Roerich
4    Danforth
Name: family, dtype: object


In [82]:
dfSurvey = pd.read_sql_query("SELECT quant FROM Survey", conexao)
print(dfSurvey)

   quant
0    rad
1    sal
2    rad
3    sal
4    rad
5    sal
6   temp
7    rad
8    sal
9   temp
10   rad
11  temp
12   sal
13   rad
14   sal
15  temp
16   sal
17   rad
18   sal
19   sal
20   rad


In [83]:
#distinct - opção para remover registros com valores repetidos de uma coluna específica
dfSurvey = pd.read_sql_query("SELECT DISTINCT quant FROM Survey;", conexao)
print(dfSurvey)

dfSurvey = pd.read_sql_query("SELECT * FROM Survey;", conexao)
print(dfSurvey['quant'])
print(dfSurvey['quant'].unique())

  quant
0   rad
1   sal
2  temp
0      rad
1      sal
2      rad
3      sal
4      rad
5      sal
6     temp
7      rad
8      sal
9     temp
10     rad
11    temp
12     sal
13     rad
14     sal
15    temp
16     sal
17     rad
18     sal
19     sal
20     rad
Name: quant, dtype: object
['rad' 'sal' 'temp']


In [84]:
dfSurvey = pd.read_sql_query("SELECT DISTINCT taken, quant FROM Survey;", conexao)
print(dfSurvey)

    taken quant
0     619   rad
1     619   sal
2     622   rad
3     622   sal
4     734   rad
5     734   sal
6     734  temp
7     735   rad
8     735   sal
9     735  temp
10    751   rad
11    751  temp
12    751   sal
13    752   rad
14    752   sal
15    752  temp
16    837   rad
17    837   sal
18    844   rad


In [85]:
#ordenando
dfPerson = pd.read_sql_query("SELECT * FROM person ORDER BY id  ASC;", conexao)
print(dfPerson)

         id   personal    family
0  danforth      Frank  Danforth
1      dyer    William      Dyer
2      lake   Anderson      Lake
3        pb      Frank   Pabodie
4       roe  Valentina   Roerich


In [86]:
#ordenando com Pandas
dfPerson = pd.read_sql_query("SELECT * FROM person", conexao)
print(dfPerson)
print(dfPerson.sort_values(by=['id']))

         id   personal    family
0      dyer    William      Dyer
1        pb      Frank   Pabodie
2      lake   Anderson      Lake
3       roe  Valentina   Roerich
4  danforth      Frank  Danforth
         id   personal    family
4  danforth      Frank  Danforth
0      dyer    William      Dyer
2      lake   Anderson      Lake
1        pb      Frank   Pabodie
3       roe  Valentina   Roerich


In [87]:
#Filtro por linhas
dfSite = pd.read_sql_query("SELECT * FROM Visited", conexao)
print(dfSite)

dfSite = pd.read_sql_query("SELECT * FROM Visited WHERE site = 'DR-1';", conexao)
print(dfSite)

dfSite = pd.read_sql_query("SELECT * FROM Visited WHERE site = 'DR-1' AND dated < '1930-01-01';", conexao)
print(dfSite)

dfSurvey= pd.read_sql_query("SELECT * FROM Survey WHERE person IN ('lake', 'roe');", conexao)
print(dfSurvey)
print(dfSite.info())

    id   site       dated
0  619   DR-1  1927-02-08
1  622   DR-1  1927-02-10
2  734   DR-3  1930-01-07
3  735   DR-3  1930-01-12
4  751   DR-3  1930-02-26
5  752   DR-3        None
6  837  MSK-4  1932-01-14
7  844   DR-1  1932-03-22
    id  site       dated
0  619  DR-1  1927-02-08
1  622  DR-1  1927-02-10
2  844  DR-1  1932-03-22
    id  site       dated
0  619  DR-1  1927-02-08
1  622  DR-1  1927-02-10
   taken person quant  reading
0    734   lake   sal     0.05
1    751   lake   sal     0.10
2    752   lake   rad     2.19
3    752   lake   sal     0.09
4    752   lake  temp   -16.00
5    752    roe   sal    41.60
6    837   lake   rad     1.46
7    837   lake   sal     0.21
8    837    roe   sal    22.50
9    844    roe   rad    11.25
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   id      2 non-null      int64 
 1   site    2 non-null      object
 2   d

In [88]:
#faz a mesma coisa que o anterior, mas pelo pandas
dfSite = pd.read_sql_query("SELECT * FROM Visited", conexao)
dfs2 = dfSite[dfSite['site'] == 'DR-1']

print(dfs2)

print(dfs2.iloc[0:-1])
#site = 'DR-1'
#print(dfSite)


    id  site       dated
0  619  DR-1  1927-02-08
1  622  DR-1  1927-02-10
7  844  DR-1  1932-03-22
    id  site       dated
0  619  DR-1  1927-02-08
1  622  DR-1  1927-02-10


In [89]:
dfSite = pd.read_sql_query("SELECT * FROM Visited;",conexao)
#dfs2 = ((dfSite['site'] == 'DR-1') & (dfSite['dated'] <= '1932-03-22'))
dfs3 = dfSite[ (dfSite['site'] == 'DR-1') ]
dfs4 = dfs3[( dfs3['dated'] == '1932-03-22')]
print(dfs4)
#dated < '1930-01-01'

    id  site       dated
7  844  DR-1  1932-03-22


               id
count    9.000000
mean   765.888889
std    117.348247
min    619.000000
25%    734.000000
50%    751.000000
75%    837.000000
max    999.000000
    id   site       dated
0  619   DR-1  1927-02-08
1  622   DR-1  1927-02-10
2  734   DR-3  1930-01-07
3  735   DR-3  1930-01-12
4  751   DR-3  1930-02-26
5  752   DR-3        None
6  837  MSK-4  1932-01-14
7  844   DR-1  1932-03-22
8  999   DDF3            
<pandas.core.groupby.generic.DataFrameGroupBy object at 0x7f7e306b4710>
site  DDF3  dated      id  site dated
8  999  DDF3      
site  DR-1  dated      id  site       dated
0  619  DR-1  1927-02-08
1  622  DR-1  1927-02-10
7  844  DR-1  1932-03-22
site  DR-3  dated      id  site       dated
2  734  DR-3  1930-01-07
3  735  DR-3  1930-01-12
4  751  DR-3  1930-02-26
5  752  DR-3        None
site  MSK-4  dated      id   site       dated
6  837  MSK-4  1932-01-14
       id  dated
site            
DDF3    1      1
DR-1    3      3
DR-3    4      3
MSK-4   1      1
site  id
site  dated
            id  site
dated               
             1     1
1927-02-08   1     1
1927-02-10   1     1
1930-01-07   1     1
1930-01-12   1     1
1930-02-26   1     1
1932-01-14   1     1
1932-03-22   1     1
dated  id
dated  site

In [90]:
dfSite = pd.read_sql_query("SELECT * FROM Visited", conexao)
dfs5 = dfSite[dfSite['site'] == "DR-3"]
dfs6 = dfs5[dfs5['dated'] >= '1930-01-10']
dfs7 = dfSite[(dfSite['site'] == "DR-3") & (dfs5['dated'] >= '1930-01-10')]
print(dfs6)
print(dfs7)

    id  site       dated
3  735  DR-3  1930-01-12
4  751  DR-3  1930-02-26
    id  site       dated
3  735  DR-3  1930-01-12
4  751  DR-3  1930-02-26


In [97]:
dfSite = pd.read_sql_query("SELECT * FROM Site", conexao)
print(dfSite)
dfvisited = pd.read_sql_query("SELECT * FROM Visited", conexao)
print(dfvisited)

    name    lat    long
0   DR-1 -49.85 -128.57
1   DR-3 -47.15 -126.72
2  MSK-4 -48.87 -123.40
    id   site       dated
0  619   DR-1  1927-02-08
1  622   DR-1  1927-02-10
2  734   DR-3  1930-01-07
3  735   DR-3  1930-01-12
4  751   DR-3  1930-02-26
5  752   DR-3        None
6  837  MSK-4  1932-01-14
7  844   DR-1  1932-03-22


In [92]:


#Exemplos com merge
dfVisitedSiteMerged = pd.read_sql_query("SELECT * FROM Site s INNER JOIN Visited v ON s.name = v.site", conexao)
print(dfVisitedSiteMerged)

dfVisitedSiteMerged = pd.read_sql_query("SELECT * FROM Site s LEFT JOIN Visited v ON s.name = v.site", conexao)
print(dfVisitedSiteMerged)

dfVisitedSiteMerged = pd.read_sql_query("SELECT s.name, s.lat FROM Site s INNER JOIN Visited v ON s.name = v.site", conexao)
print(dfVisitedSiteMerged)

    name    lat    long   id   site       dated
0   DR-1 -49.85 -128.57  619   DR-1  1927-02-08
1   DR-1 -49.85 -128.57  622   DR-1  1927-02-10
2   DR-1 -49.85 -128.57  844   DR-1  1932-03-22
3   DR-3 -47.15 -126.72  734   DR-3  1930-01-07
4   DR-3 -47.15 -126.72  735   DR-3  1930-01-12
5   DR-3 -47.15 -126.72  751   DR-3  1930-02-26
6   DR-3 -47.15 -126.72  752   DR-3        None
7  MSK-4 -48.87 -123.40  837  MSK-4  1932-01-14
    name    lat    long   id   site       dated
0   DR-1 -49.85 -128.57  619   DR-1  1927-02-08
1   DR-1 -49.85 -128.57  622   DR-1  1927-02-10
2   DR-1 -49.85 -128.57  844   DR-1  1932-03-22
3   DR-3 -47.15 -126.72  734   DR-3  1930-01-07
4   DR-3 -47.15 -126.72  735   DR-3  1930-01-12
5   DR-3 -47.15 -126.72  751   DR-3  1930-02-26
6   DR-3 -47.15 -126.72  752   DR-3        None
7  MSK-4 -48.87 -123.40  837  MSK-4  1932-01-14
    name    lat
0   DR-1 -49.85
1   DR-1 -49.85
2   DR-1 -49.85
3   DR-3 -47.15
4   DR-3 -47.15
5   DR-3 -47.15
6   DR-3 -47.15
7  MSK-4

In [93]:
#dfMerge = pd.merge(dfsite, dfvisited)
#print(dfMerge)

dfMerge = pd.merge(dfsite, dfvisited, left_on='name',right_on='site')
print(dfMerge)

dfMerge = pd.merge(dfsite, dfvisited, left_on='name',right_on='site').drop('site', axis=1)
print(dfMerge)

    name    lat    long   id   site       dated
0   DR-1 -49.85 -128.57  619   DR-1  1927-02-08
1   DR-1 -49.85 -128.57  622   DR-1  1927-02-10
2   DR-1 -49.85 -128.57  844   DR-1  1932-03-22
3   DR-3 -47.15 -126.72  734   DR-3  1930-01-07
4   DR-3 -47.15 -126.72  735   DR-3  1930-01-12
5   DR-3 -47.15 -126.72  751   DR-3  1930-02-26
6   DR-3 -47.15 -126.72  752   DR-3        None
7  MSK-4 -48.87 -123.40  837  MSK-4  1932-01-14
    name    lat    long   id       dated
0   DR-1 -49.85 -128.57  619  1927-02-08
1   DR-1 -49.85 -128.57  622  1927-02-10
2   DR-1 -49.85 -128.57  844  1932-03-22
3   DR-3 -47.15 -126.72  734  1930-01-07
4   DR-3 -47.15 -126.72  735  1930-01-12
5   DR-3 -47.15 -126.72  751  1930-02-26
6   DR-3 -47.15 -126.72  752        None
7  MSK-4 -48.87 -123.40  837  1932-01-14


In [98]:
dfMerge = pd.merge(dfsite, dfvisited, left_on='name',right_on='site', how='inner').drop('site', axis=1)
print(dfMerge)

dfMerge = pd.merge(dfsite, dfvisited, left_on='name',right_on='site', how='outer').drop('site', axis=1)
print(dfMerge)

dfMerge = pd.merge(dfsite, dfvisited, left_on='name',right_on='site', how='left').drop('site', axis=1)
print(dfMerge)

dfMerge = pd.merge(dfsite, dfvisited, left_on='name',right_on='site', how='right').drop('site', axis=1)
print(dfMerge)

    name    lat    long   id       dated
0   DR-1 -49.85 -128.57  619  1927-02-08
1   DR-1 -49.85 -128.57  622  1927-02-10
2   DR-1 -49.85 -128.57  844  1932-03-22
3   DR-3 -47.15 -126.72  734  1930-01-07
4   DR-3 -47.15 -126.72  735  1930-01-12
5   DR-3 -47.15 -126.72  751  1930-02-26
6   DR-3 -47.15 -126.72  752        None
7  MSK-4 -48.87 -123.40  837  1932-01-14
    name    lat    long   id       dated
0   DR-1 -49.85 -128.57  619  1927-02-08
1   DR-1 -49.85 -128.57  622  1927-02-10
2   DR-1 -49.85 -128.57  844  1932-03-22
3   DR-3 -47.15 -126.72  734  1930-01-07
4   DR-3 -47.15 -126.72  735  1930-01-12
5   DR-3 -47.15 -126.72  751  1930-02-26
6   DR-3 -47.15 -126.72  752        None
7  MSK-4 -48.87 -123.40  837  1932-01-14
    name    lat    long   id       dated
0   DR-1 -49.85 -128.57  619  1927-02-08
1   DR-1 -49.85 -128.57  622  1927-02-10
2   DR-1 -49.85 -128.57  844  1932-03-22
3   DR-3 -47.15 -126.72  734  1930-01-07
4   DR-3 -47.15 -126.72  735  1930-01-12
5   DR-3 -47.15 



Titanic

1 Leia os dados train.csv usando pandas

2 Remova a coluna com nomes de pessoas

3 A coluna Cabin contém uma letra e um número. Crie uma nova coluna chamada deck usando apenas a letra de cada registro presente na coluna Cabin.

4 Substitua os valores ausentes pela média de valores da coluna

5 Os valores da coluna "embarked" sao os seguintes: C = Cherbourg, Q = Queenstown, S = Southampton

6 Crie um CSV com essas uma coluna nome_cidade_de_embarque: C = Cherbourg, Q = Queenstown, S = Southampton e mais uma coluna chamada população com valores hipotéticos, faça um merge entre o CSV train.csv e esse novo CSV com os nomes das cidades. Implemente um filtro de passageiros por população da cidade de embarque


In [10]:
###EXERCICIO TITANIC
#1
#fromcsv = pd.read_csv('/home/fernando/github/ResidenciaIA_Senai/aula_sql/train.csv')
fromcsv = pd.read_csv('/Users/naka/Documents/GitHub/ResidenciaIA_Senai/aula_sql/train.csv')
#2
fromcsv.head().drop('Name', axis = 1)
print(fromcsv)





     PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ... 

In [11]:

#3
deck = fromcsv['Cabin']
deck = deck.astype(str).str[0]
#print(deck)
fromcsv['Deck'] = deck
print(fromcsv)

     PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ... 

In [27]:
#4
import numpy as np

age = fromcsv['Age']
#age = pd.Series([1,2,3,np.nan,4])
age_mean = age.mean()
print(age_mean)
values = {'Age': age_mean}
fromcsv = fromcsv.fillna(value = values)
print(fromcsv)

29.69911764705882
     PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex        Age  \
0                              Braund, Mr. Owen Harris    male  22.000000   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.000000   
2                               Heikkinen, Miss. Laina  female  26.000000   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.000000   
4                             Allen, Mr. William Henry    male  35.000000   
..                                                 ...     ... 

In [30]:
#5
#C = Cherbourg, Q = Queenstown, S = Southampton - nome_cidade_embarque
cidades_csv = pd.read_csv('/Users/naka/Documents/GitHub/ResidenciaIA_Senai/aula_sql/cidades.csv')
print(cidades_csv)
            

  nome_cidade_embarque   populacao
0            Cherbourg       55000
1           Queenstown     1345660
2          Southampton      300000
